src/Microsoft.AspNet.OData.Shared/Query/Expressions/SelectExpandBinder.cs

```diff
@@ -428,7 +342,225 @@ private Expression ProjectElement(Expression source, SelectExpandClause selectEx
+        /// Process the <see cref="PathSelectItem"/>.
+        /// </summary>
+        /// <param name="pathSelectItem">The selected item.</param>
+        /// <param name="navigationSource">The navigation source.</param>
+        /// <param name="currentLevelPropertiesInclude">The current level properties included.</param>
+        /// <returns>true if it's dynamic property selection, false if it's not.</returns>
+        private static bool ProcessSelectedItem(PathSelectItem pathSelectItem,
+            IEdmNavigationSource navigationSource,
+            IDictionary<IEdmStructuralProperty, SelectExpandIncludeProperty> currentLevelPropertiesInclude)
+        {
+            Contract.Assert(pathSelectItem != null && pathSelectItem.SelectedPath != null);
+            Contract.Assert(currentLevelPropertiesInclude != null);
+
+            // Verify and process the $select path
+            IList<ODataPathSegment> remainingSegments;
+            ODataPathSegment firstNonTypeSegment = pathSelectItem.SelectedPath.GetFirstNonTypeCastSegment(out remainingSegments);
+
+            // for $select=NS.SubType/Property, we don't care about the leading type segment, because with or without the type segment
+            // the "Property" property value expression should be built into the property container.
+
+            PropertySegment firstSturucturalPropertySegment = firstNonTypeSegment as PropertySegment;
+            if (firstSturucturalPropertySegment != null)
+            {
+                // $select=abc/..../xyz
+                SelectExpandIncludeProperty newPropertySelectItem;
+                if (!currentLevelPropertiesInclude.TryGetValue(firstSturucturalPropertySegment.Property, out newPropertySelectItem))
+                {
+                    newPropertySelectItem = new SelectExpandIncludeProperty(firstSturucturalPropertySegment, navigationSource);
+                    currentLevelPropertiesInclude[firstSturucturalPropertySegment.Property] = newPropertySelectItem;
+                }
+
+                newPropertySelectItem.AddSubSelectItem(remainingSegments, pathSelectItem);
+            }
+            else
+            {
+                // If we can't find a PropertySegment, the $select path maybe selecting an operation, a navigation or dynamic property.
+                // And the remaing segments should be null.
+                Contract.Assert(remainingSegments == null);
+
+                // For operation (action/function), needn't process it.
+                // For navigation property, needn't process it here.
+
+                // For dynamic property, let's test the last segment for this path select item.
+                if (firstNonTypeSegment is DynamicPathSegment)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static bool IsSelectsOpenTypeSegments(SelectExpandClause selectExpandClause, IEdmStructuredType structuredType)
         {
```

Comment:
Can you add a comment describing what this function is intended to do (and perhaps a more intuitive name -- IsSelects... is a bit strange...)? It looks like the function changed behavior, and I can't tell whether the new behavior is correct. Should it return whether all dynamic properties should be returned (which, I think, is the current behavior, and could be named something like "IncludesAllDynamicProperties") or whether there are any dynamic properties explicitly selected (which, I think, was the previous behavior, and would fit "SelectIncludesDynamicProperties").

Reply:
The change here is for 
1)  algorithm changed, now we consider the select/expand step by step, we can't test the whole "path" to see whether there's a dynamic property or not.

2) performance consideration, we don't need to travel the "SelectedItems" again and again here, because in the previous logic, we travel the "SelectItems" one by one and test each Select item

And yes, the function name is not correct. So, I changed to "IsSelectAllOnOpenType".




---
In reply to: [355543290](https://github.com/OData/WebApi/pull/1941#discussion_r355543290) [](ancestors = 355543290)